File: frontend/src/types/investigation.ts
export type ThreatSeverity = 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW' | 'CLEAN';

export type CaseStatus =
  | 'NEW'
  | 'TRIAGING'
  | 'INVESTIGATING'
  | 'CONTAINED'
  | 'RESOLVED'
  | 'CLOSED'
  | 'FALSE_POSITIVE'
  | 'OPEN'
  | 'IN_REVIEW';

export type ThreatClassification =
  | 'PHISHING'
  | 'BUSINESS_EMAIL_COMPROMISE'
  | 'FINANCIAL_FRAUD'
  | 'CREDENTIAL_HARVESTING'
  | 'MALICIOUS_ATTACHMENT'
  | 'MALWARE'
  | 'SUSPICIOUS'
  | 'BENIGN'
  | 'IMPERSONATION'
  | 'UNKNOWN';

export type PriorityLevel = 'P1_CRITICAL' | 'P2_HIGH' | 'P3_MEDIUM' | 'P4_LOW';

export type NoteType = 'NOTE' | 'ACTION' | 'DECISION' | 'ESCALATION' | 'SYSTEM';

export interface AnalystNote {
  note_id: string;
  investigation_id: string;
  author: string;
  timestamp: string;
  content: string;
  note_type: NoteType;
}

export interface InvestigationSummary {
  id: string;
  case_number?: string;
  title: string;
  sender: string;
  subject: string;
  severity: ThreatSeverity;
  threat_score: number;
  risk_score?: number;
  classification?: string;
  confidence?: number;
  status: string;
  analyst: string;
  assigned_analyst?: string;
  priority?: string;
  tags?: string[];
  evidence_hash: string;
  blockchain_verified: boolean;
  blockchain_status?: string;
  created_at: string;
  updated_at: string;
}

export interface Investigation {
  id: string;
  case_number: string;
  title: string;
  description: string;
  source: string;
  created_at: string;
  updated_at: string;
  status: string;
  severity: ThreatSeverity;
  classification: string;
  risk_score: number;
  threat_score?: number;
  confidence: number;
  assigned_analyst: string;
  analyst?: string;
  tags: string[];
  priority: PriorityLevel;
  email_evidence_id?: string;
  evidence_id?: string;
  evidence_hash: string;
  sender: string;
  subject: string;
  blockchain_status: string;
  blockchain_tx?: string;
  blockchain_block?: number;
  blockchain_network?: string;
  blockchain_verified: boolean;
  created_by: string;
  notes: AnalystNote[];
  verdict_summary?: string;
  recommended_actions: string[];
}

export interface InvestigationListResponse {
  total: number;
  investigations: InvestigationSummary[];
}

export interface InvestigationOverview {
  investigation: Investigation;
  top_indicators: Array<{
    name: string;
    category: string;
    severity: string;
    score_impact: number;
    description: string;
    evidence_excerpt: string;
  }>;
  top_entities: {
    ips_count: number;
    domains_count: number;
    urls_count: number;
    attachments_count: number;
    suspicious_domains: string[];
    suspicious_urls: string[];
    attachments: string[];
  };
  auth_summary: {
    spf?: string;
    dkim?: string;
    dmarc?: string;
  };
  key_findings: string[];
  blockchain_summary: {
    status: string;
    verified: boolean;
    evidence_id?: string;
    evidence_hash?: string;
    transaction_hash?: string;
    block_number?: number;
    network?: string;
  };
  latest_activity: Array<{
    timestamp: string;
    author: string;
    type: string;
    content: string;
  }>;
}

export interface DashboardStats {
  total_investigations: number;
  critical_cases: number;
  high_cases: number;
  medium_cases: number;
  low_clean_cases: number;
  cases_today: number;
  blockchain_verified_count: number;
  blockchain_verified_percentage: number;
  tamper_alerts: number;
  mean_latency_seconds: number;
  classification_breakdown: Record<string, number>;
  status_breakdown: Record<string, number>;
  top_source_ips: Array<{
    ip: string;
    count: number;
    country: string;
    threat: string;
  }>;
  top_domains: Array<{
    domain: string;
    count: number;
    type: string;
    risk: string;
  }>;
  top_countries: Array<{
    country_code: string;
    country_name: string;
    incidents: number;
    percentage: number;
  }>;
}

export interface RunFullInvestigationResult {
  investigation: Investigation;
  steps_completed: string[];
  verdict: {
    classification: string;
    severity: string;
    risk_score: number;
    confidence: number;
    evidence_hash: string;
    blockchain_block?: number;
    blockchain_tx?: string;
    integrity_verified: boolean;
  };
}

export interface InvestigationReportData {
  case: Investigation;
  forensic_summary: any;
  threat_assessment: any;
  intelligence_summary: any;
  timeline_events: any[];
  custody_chain: any[];
  exported_at: string;
  report_id: string;
}

File: backend/app/schemas/investigation.py
from datetime import datetime
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field

from app.services.investigations.models import (
    InvestigationStatus,
    InvestigationSeverity,
    InvestigationClassification,
    InvestigationPriority,
    NoteType,
    AnalystNote,
    Investigation,
    InvestigationOverview,
)


class InvestigationSummary(BaseModel):
    id: str = Field(..., description="Investigation ID, e.g. INV-2026-00001")
    case_number: Optional[str] = Field(default=None, description="Internal case reference")
    title: str = Field(..., description="Case Title")
    sender: str = Field(..., description="Suspect Sender Address")
    subject: str = Field(..., description="Email Subject")
    severity: str = Field(..., description="Severity level: CRITICAL, HIGH, MEDIUM, LOW, CLEAN")
    threat_score: int = Field(..., ge=0, le=100, description="Calculated threat risk score (0-100)")
    risk_score: Optional[int] = Field(default=None, description="Alias for threat_score")
    classification: Optional[str] = Field(default=None, description="Threat Classification")
    confidence: Optional[int] = Field(default=85, description="Model/Rule confidence percentage")
    status: str = Field(..., description="Investigation status: NEW, TRIAGING, INVESTIGATING, etc.")
    analyst: str = Field(..., description="Assigned SOC Analyst")
    assigned_analyst: Optional[str] = Field(default=None, description="Alias for analyst")
    priority: Optional[str] = Field(default="P3_MEDIUM", description="Case Priority")
    tags: List[str] = Field(default_factory=list)
    evidence_hash: str = Field(..., description="SHA-256 hash of core evidence")
    blockchain_verified: bool = Field(default=False, description="Verification status on blockchain ledger")
    blockchain_status: Optional[str] = Field(default="PENDING")
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class InvestigationListResponse(BaseModel):
    total: int
    investigations: List[InvestigationSummary]


class CreateInvestigationRequest(BaseModel):
    title: str = Field(..., description="Case Title")
    description: Optional[str] = Field(default="")
    source: Optional[str] = Field(default="SOC_MANUAL")
    severity: Optional[InvestigationSeverity] = Field(default=InvestigationSeverity.MEDIUM)
    classification: Optional[InvestigationClassification] = Field(default=InvestigationClassification.SUSPICIOUS)
    risk_score: Optional[int] = Field(default=0, ge=0, le=100)
    assigned_analyst: Optional[str] = Field(default="UNASSIGNED")
    priority: Optional[InvestigationPriority] = Field(default=InvestigationPriority.P3_MEDIUM)
    tags: Optional[List[str]] = Field(default_factory=list)
    sender: Optional[str] = Field(default="")
    subject: Optional[str] = Field(default="")
    evidence_hash: Optional[str] = Field(default="")


class UpdateInvestigationRequest(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    status: Optional[InvestigationStatus] = None
    severity: Optional[InvestigationSeverity] = None
    classification: Optional[InvestigationClassification] = None
    risk_score: Optional[int] = Field(default=None, ge=0, le=100)
    assigned_analyst: Optional[str] = None
    priority: Optional[InvestigationPriority] = None
    tags: Optional[List[str]] = None
    verdict_summary: Optional[str] = None


class AddNoteRequest(BaseModel):
    author: Optional[str] = Field(default="SOC Analyst")
    content: str = Field(..., description="Note text or decision")
    note_type: Optional[NoteType] = Field(default=NoteType.NOTE)


class AssignAnalystRequest(BaseModel):
    analyst: str = Field(..., description="Analyst name or ID to assign")


class UpdateStatusRequest(BaseModel):
    status: InvestigationStatus = Field(..., description="New investigation status")
    reason: Optional[str] = Field(default="Status transition updated by analyst")
    analyst: Optional[str] = Field(default="SOC Analyst")


class UpdateTagsRequest(BaseModel):
    tags: List[str] = Field(..., description="List of tags to assign")


class EscalateRequest(BaseModel):
    reason: str = Field(..., description="Justification for escalation")
    escalate_to: Optional[str] = Field(default="TIER_3_INCIDENT_RESPONSE")
    analyst: Optional[str] = Field(default="SOC Analyst")


class FalsePositiveRequest(BaseModel):
    reason: str = Field(..., description="Reasoning for false positive classification")
    analyst: Optional[str] = Field(default="SOC Analyst")


class RunFullInvestigationResponse(BaseModel):
    investigation: Investigation
    steps_completed: List[str]
    verdict: Dict[str, Any]


class DashboardStatsResponse(BaseModel):
    total_investigations: int
    critical_cases: int
    high_cases: int
    medium_cases: int
    low_clean_cases: int
    cases_today: int
    blockchain_verified_count: int
    blockchain_verified_percentage: float
    tamper_alerts: int
    mean_latency_seconds: float
    classification_breakdown: Dict[str, int]
    status_breakdown: Dict[str, int]
    top_source_ips: List[Dict[str, Any]]
    top_domains: List[Dict[str, Any]]
    top_countries: List[Dict[str, Any]]
